File: java/solitaire/game/src/main/java/org/silnith/game/GameState.java
package org.silnith.game;

import org.silnith.game.move.Move;
import org.silnith.game.util.Pair;

/**
 * A single node in a game tree.
 * 
 * @param <M> the move type for the game. This can simply be {@code Move<B>},
 *            but the interface allows specifying a subtype in the case that a
 *            game-specific interface is needed.
 * @param <B> the board type for the game
 * @author <a href="mailto:silnith@gmail.com">Kent Rosenkoetter</a>
 */
public class GameState<M extends Move<B>, B> extends Pair<M, B> {

    /**
     * Constructs a game state with the given move and board.
     *
     * @param move  the move
     * @param board the board
     */
    public GameState(final M move, final B board) {
        super(move, board);
    }

    /**
     * Returns the move.
     *
     * @return the move
     */
    public M getMove() {
        return getFirst();
    }

    /**
     * Returns the board.
     *
     * @return the board
     */
    public B getBoard() {
        return getSecond();
    }

}

File: csharp/Solitaire/Game/GameState.cs
using System;

namespace Silnith.Game
{
    /// <summary>
    /// A single node in a game tree.
    /// </summary>
    /// <typeparam name="M">The type of the game moves.  This can simply be <see cref="IMove{B}"/>,
    /// but the interface allows specifying a subtype in the case that a game-specific interface is needed.</typeparam>
    /// <typeparam name="B">The type of the game board.</typeparam>
    public class GameState<M, B> : Tuple<M, B>
        where M : IMove<B>
    {
        /// <summary>
        /// The move.
        /// </summary>
        public virtual M Move
        {
            get
            {
                return Item1;
            }
        }

        /// <summary>
        /// The board.
        /// </summary>
        public virtual B Board
        {
            get
            {
                return Item2;
            }
        }

        /// <summary>
        /// Constructs a game state with the given move and board.
        /// </summary>
        /// <param name="initialMove">The move.</param>
        /// <param name="initialBoard">The board.</param>
        public GameState(M initialMove, B initialBoard) : base(initialMove, initialBoard)
        {
        }
    }
}
